File: javatests/com/google/fhir/common/ResourceUtilsTest.java
//    Copyright 2021 Google Inc.
//
//    Licensed under the Apache License, Version 2.0 (the "License");
//    you may not use this file except in compliance with the License.
//    You may obtain a copy of the License at
//
//        https://www.apache.org/licenses/LICENSE-2.0
//
//    Unless required by applicable law or agreed to in writing, software
//    distributed under the License is distributed on an "AS IS" BASIS,
//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//    See the License for the specific language governing permissions and
//    limitations under the License.

package com.google.fhir.common;

import static com.google.common.truth.Truth.assertThat;
import static com.google.common.truth.extensions.proto.ProtoTruth.assertThat;
import static com.google.fhir.common.ProtoUtils.findField;
import static java.nio.charset.StandardCharsets.UTF_8;
import static org.junit.Assert.assertThrows;

import com.google.common.collect.ImmutableCollection;
import com.google.common.collect.ImmutableList;
import com.google.common.io.Files;
import com.google.devtools.build.runfiles.Runfiles;
import com.google.protobuf.Message;
import com.google.protobuf.TextFormat;
import java.io.File;
import java.io.IOException;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.junit.runners.Parameterized;

@RunWith(Parameterized.class)
public final class ResourceUtilsTest {
  private final Message bundleType;
  private final Message referenceType;

  private Runfiles runfiles;

  public ResourceUtilsTest(Message bundleType, Message referenceType) {
    this.bundleType = bundleType;
    this.referenceType = referenceType;
  }

  @Before
  public void setUp() throws IOException {
    runfiles = Runfiles.create();
  }

  @Parameterized.Parameters
  public static ImmutableCollection<Object[]> params() {
    return ImmutableList.of(
        new Object[] {
          com.google.fhir.r4.core.Bundle.getDefaultInstance(),
          com.google.fhir.r4.core.Reference.getDefaultInstance()
        },
        new Object[] {
          com.google.fhir.stu3.proto.Bundle.getDefaultInstance(),
          com.google.fhir.stu3.proto.Reference.getDefaultInstance()
        });
  }

  private static Message.Builder setStringField(
      Message.Builder builder, String field, String stringValue) {
    Message.Builder fieldBuilder = builder.getFieldBuilder(findField(builder, field));
    fieldBuilder.setField(findField(fieldBuilder, "value"), stringValue);
    return builder;
  }

  /** Read the specifed prototxt file from the testdata directory and parse it. */
  private void mergeText(String filename, Message.Builder builder) throws IOException {
    File file = new File(runfiles.rlocation("com_google_fhir/testdata/" + filename));
    TextFormat.merge(Files.asCharSource(file, UTF_8).read(), builder);
  }

  @Test
  public void testGetContainedResource_hasContained() {
    Message.Builder entryBuilder =
        bundleType.newBuilderForType().newBuilderForField(findField(bundleType, "entry"));
    Message.Builder containedResourceBuilder =
        entryBuilder.newBuilderForField(findField(entryBuilder, "resource"));
    Message.Builder conceptMap =
        containedResourceBuilder.getFieldBuilder(
            findField(containedResourceBuilder, "concept_map"));
    setStringField(conceptMap, "id", "my-concept-map");

    assertThat(ResourceUtils.getContainedResource(containedResourceBuilder.build()).get())
        .isEqualTo(conceptMap.build());
  }

  @Test
  public void testGetContainedResource_noContained() {
    Message.Builder entryBuilder =
        bundleType.newBuilderForType().newBuilderForField(findField(bundleType, "entry"));
    Message.Builder containedResourceBuilder =
        entryBuilder.newBuilderForField(findField(entryBuilder, "resource"));

    assertThat(ResourceUtils.getContainedResource(containedResourceBuilder.build()).isPresent())
        .isFalse();
  }

  @Test
  public void testGetContainedResource_notContainedThrows() {
    assertThrows(
        IllegalArgumentException.class,
        () -> ResourceUtils.getContainedResource(bundleType.getDefaultInstanceForType()));
  }

  @Test
  public void testSplitIfRelativeReference_referenceWithoutUriUnchanged() {
    Message.Builder reference = referenceType.newBuilderForType();
    setStringField(reference, "fragment", "arbitrary-fragment");

    Message expected = reference.build();
    ResourceUtils.splitIfRelativeReference(reference);
    assertThat(reference.build()).isEqualTo(expected);
  }

  @Test
  public void testSplitIfRelativeReference_fragmentMovedToFragmentField() {
    Message.Builder reference = referenceType.newBuilderForType();
    setStringField(reference, "uri", "#someFragment");

    Message.Builder expected = referenceType.newBuilderForType();
    setStringField(expected, "fragment", "someFragment");

    ResourceUtils.splitIfRelativeReference(reference);
    assertThat(reference.build()).isEqualTo(expected.build());
  }

  @Test
  public void testSplitIfRelativeReference_relativeReferenceSlotted_withoutHistory() {
    Message.Builder reference = referenceType.newBuilderForType();
    setStringField(reference, "uri", "ActivityDefinition/ABCD");

    Message.Builder expected = referenceType.newBuilderForType();
    setStringField(expected, "activity_definition_id", "ABCD");

    ResourceUtils.splitIfRelativeReference(reference);
    assertThat(reference.build()).isEqualTo(expected.build());
  }

  @Test
  public void testSplitIfRelativeReference_relativeReferenceSlotted_withHistory() {
    Message.Builder reference = referenceType.newBuilderForType();
    setStringField(reference, "uri", "ActivityDefinition/ABCD/_history/some_version_id");

    Message.Builder expected = referenceType.newBuilderForType();
    Message.Builder referenceId =
        expected.getFieldBuilder(findField(expected, "activity_definition_id"));
    referenceId.setField(findField(referenceId, "value"), "ABCD");
    setStringField(referenceId, "history", "some_version_id");

    ResourceUtils.splitIfRelativeReference(reference);
    assertThat(reference.build()).isEqualTo(expected.build());
  }

  @Test
  public void testResolveBundleReferences() throws IOException {
    Message.Builder bundle = bundleType.newBuilderForType();
    mergeText("resource_utils-resolveBundleReferences-input.prototxt", bundle);

    ResourceUtils.resolveBundleReferences(bundle);

    Message.Builder expected = bundleType.newBuilderForType();
    mergeText("resource_utils-resolveBundleReferences-output.prototxt", expected);

    assertThat(bundle.build()).isEqualTo(expected.build());
  }
}

File: py/google/fhir/utils/resource_utils_test.py
#
# Copyright 2021 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
"""Test resource_utils functionality."""

from absl.testing import absltest
from proto.google.fhir.proto.r4.core import datatypes_pb2
from proto.google.fhir.proto.r4.core.resources import account_pb2
from proto.google.fhir.proto.r4.core.resources import activity_definition_pb2
from proto.google.fhir.proto.r4.core.resources import bundle_and_contained_resource_pb2
from proto.google.fhir.proto.r4.core.resources import structure_definition_pb2
from google.fhir.utils import resource_utils

# TODO: These util tests should not be FHIR-specific.


def _bundle_with_single_structure_definition(
) -> bundle_and_contained_resource_pb2.Bundle:
  """Returns a `Bundle` with a single `StructureDefinition` entry."""
  return bundle_and_contained_resource_pb2.Bundle(entry=[
      bundle_and_contained_resource_pb2.Bundle.Entry(
          resource=bundle_and_contained_resource_pb2.ContainedResource(
              structure_definition=structure_definition_pb2.StructureDefinition(
                  name=datatypes_pb2.String(value='Test'))))
  ])


class ResourceUtilsTest(absltest.TestCase):
  """Performs unit tests targeting `resource_utils`."""

  def testExtractResourcesFromBundle_withSetResource_succeeds(self):
    bundle = _bundle_with_single_structure_definition()
    actual = resource_utils.extract_resources_from_bundle(
        bundle, resource_type=structure_definition_pb2.StructureDefinition)
    self.assertLen(actual, 1)

  def testExtractResourcesFromBundle_withUnsetResource_returnsEmpty(self):
    bundle = _bundle_with_single_structure_definition()
    actual = resource_utils.extract_resources_from_bundle(
        bundle, resource_type=account_pb2.Account)
    self.assertEmpty(actual)

  def testExtractResourceFromBundle_withInvalidResource_raises(self):
    bundle = _bundle_with_single_structure_definition()
    with self.assertRaises(ValueError) as ve:
      # Use a nested message that has no mapping to ContainedResource oneof
      _ = resource_utils.extract_resources_from_bundle(
          bundle, resource_type=account_pb2.Account.StatusCode)
    self.assertIsInstance(ve.exception, ValueError)

  def testExtractResourceFromBundle_withNonBundle_raises(self):
    non_bundle = account_pb2.Account()
    with self.assertRaises(TypeError) as te:
      _ = resource_utils.extract_resources_from_bundle(
          non_bundle, resource_type=activity_definition_pb2.ActivityDefinition)
    self.assertIsInstance(te.exception, TypeError)

  def testGetContainedResource_withValidContainedResource_succeeds(self):
    expected_account = account_pb2.Account(
        name=datatypes_pb2.String(value='TestAccount'))
    contained_resource = bundle_and_contained_resource_pb2.ContainedResource(
        account=expected_account)
    actual_account = resource_utils.get_contained_resource(contained_resource)
    self.assertEqual(actual_account, expected_account)

  def testGetContainedResource_withInvalidContainedResource_raises(self):
    not_contained_resource = account_pb2.Account()
    with self.assertRaises(TypeError) as te:
      _ = resource_utils.get_contained_resource(not_contained_resource)
    self.assertIsInstance(te.exception, TypeError)


if __name__ == '__main__':
  absltest.main()
